File: bindings/nodejs/examples/how_tos/native_tokens/burn.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { getUnlockedWallet } from '../../wallet/common';

// The minimum available native token amount to search for in the wallet.
const MIN_AVAILABLE_AMOUNT = BigInt(11);
// The amount of the native token to burn.
const BURN_AMOUNT = BigInt(1);

// In this example we will burn a native token. This will not increase the melted supply in the foundry,
// therefore the foundry output is also not required. But this will also make it impossible to destroy the foundry
// output that minted it.
//
// Make sure that `STRONGHOLD_SNAPSHOT_PATH` and `WALLET_DB_PATH` already exist by
// running the `how_tos/wallet/create-wallet` example!
//
// Rename `.env.example` to `.env` first, then run
// yarn run-example ./how_tos/native_tokens/burn.ts
async function run() {
    for (const envVar of ['EXPLORER_URL']) {
        if (!(envVar in process.env)) {
            throw new Error(`.env ${envVar} is undefined, see .env.example`);
        }
    }
    try {
        // Create the wallet
        const wallet = await getUnlockedWallet();

        // May want to ensure the wallet is synced before sending a transaction.
        let balance = await wallet.sync();

        // Get a token with sufficient balance
        const tokenId = Object.keys(balance.nativeTokens).find(
            (t) => balance.nativeTokens[t].available >= MIN_AVAILABLE_AMOUNT,
        );

        if (!tokenId) {
            throw new Error(
                `No native token with at least '${Number(
                    MIN_AVAILABLE_AMOUNT,
                )}' tokens of it in the wallet`,
            );
        }
        let token = balance.nativeTokens[tokenId];

        console.log(`Balance before burning: ${token.available}`);

        // Burn a native token
        const transaction = await wallet
            .prepareBurnNativeToken(tokenId, BURN_AMOUNT)
            .then((prepared) => prepared.send());

        console.log(`Transaction sent: ${transaction.transactionId}`);

        await wallet.waitForTransactionAcceptance(transaction.transactionId);

        console.log(
            `Tx accepted: ${process.env.EXPLORER_URL}/transactions/${transaction.transactionId}`,
        );

        balance = await wallet.sync();

        token = balance.nativeTokens[tokenId];
        if (token) {
            console.log(`Balance after burning: ${token.available}`);
        } else {
            console.log(`No remaining tokens`);
        }
    } catch (error) {
        console.log('Error: ', error);
    }
    process.exit(0);
}

void run().then(() => process.exit());

File: bindings/python/examples/how_tos/native_tokens/burn.py
import os
from dotenv import load_dotenv
from iota_sdk import Wallet, WalletOptions

load_dotenv()

# In this example we will burn native tokens

wallet = Wallet(WalletOptions(storage_path=os.environ.get('WALLET_DB_PATH')))

# Sync wallet with the node
balance = wallet.sync()

if 'STRONGHOLD_PASSWORD' not in os.environ:
    raise Exception(".env STRONGHOLD_PASSWORD is undefined, see .env.example")

wallet.set_stronghold_password(os.environ["STRONGHOLD_PASSWORD"])

# Find native token with enough balance
token_id, token = next(
    (k, v) for k, v in balance.native_tokens.items() if v.available >= 10)
print(f'Balance before burning: {token.available}')

burn_amount = 1

# Send transaction.
transaction = wallet.prepare_burn_native_token(
    token_id, burn_amount).send()
print(f'Transaction sent: {transaction.transaction_id}')

wallet.wait_for_transaction_acceptance(
    transaction.transaction_id)
print(
    f'Tx accepted: {os.environ["EXPLORER_URL"]}/transactions/{transaction.transaction_id}')

balance = wallet.sync()
available_balance = balance.native_tokens[token_id].available
print(f'Balance after burning: {available_balance}')
